pyharmonysearch/harmony_search.py

```diff
@@ -126,32 +133,50 @@ def run(self):
                 else:
                     self._random_selection(harmony, i)
             fitness = self._obj_fun.get_fitness(harmony)
-
             self._update_harmony_memory(harmony, fitness)
-
             num_imp += 1
 
+            # save harmonies every n_inds th improvisations (i.e., one 'generation')
+            if num_imp % self._obj_fun.get_hms() == 0:
+                generation += 1
+                harmony_list = {'gen': generation, 'harmonies': self._harmony_memory}
+                self._harmony_history.append(harmony_list)
+
         # return best harmony
         best_harmony = None
         best_fitness = float('-inf') if self._obj_fun.maximize() else float('+inf')
         for harmony, fitness in self._harmony_memory:
             if (self._obj_fun.maximize() and fitness > best_fitness) or (not self._obj_fun.maximize() and fitness < best_fitness):
                 best_harmony = harmony
                 best_fitness = fitness
-        return best_harmony, best_fitness, self._harmony_memory
+        return best_harmony, best_fitness, self._harmony_memory, self._harmony_history
 
-    def _initialize(self):
+    def _initialize(self, initial_harmonies=None):
         """
             Initialize harmony_memory, the matrix (list of lists) containing the various harmonies (solution vectors). Note
             that we aren't actually doing any matrix operations, so a library like NumPy isn't necessary here. The matrix
             merely stores previous harmonies.
+
+            If harmonies are provided, then use them instead of randomly initializing them.
         """
+        if initial_harmonies is not None:
+            if len(initial_harmonies) != self._obj_fun.get_hms():
+                raise ValueError('Number of initial harmonies does not equal to the harmony memory size.')
```

Comment:
Modify this slightly to:

```python
raise ValueError('Number of initial harmonies ({}) does not equal the harmony memory size ({}).'.format(len(initial_harmonies), self._obj_fun.get_hms())
```

Although you should pull the variables first to avoid recomputing them.